### PR TITLE
Add otelaws middleware to support trace propagation in the AWS SDK v2 module

### PR DESCRIPTION
**Pull request by lukestoward**
Currently the `otelaws` instrumentation module does not propagate tracing information as part of any AWS SDK v2 call. To support this, I've added a new finalizer middleware that uses the provided text map propagator (or global using the OTel API) and injects the trace information into the request headers.

This functionality is really important for me as I am heavily dependent on the AWS SDK v2 Go module and, most importantly, tracing across AWS service calls.

### CODE REVIEW

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go [hunk 1]
```diff
@@ -110,6 +112,25 @@ func (m otelMiddlewares) deserializeMiddleware(stack *middleware.Stack) error {
...
+	return stack.Finalize.Add(middleware.FinalizeMiddlewareFunc("OTelFinalizeMiddleware", func(
+		ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (
+		out middleware.FinalizeOutput, metadata middleware.Metadata, err error) {
+
+		// Propagate the Trace information by injecting it into the HTTP request.
+		switch req := in.Request.(type) {
+		case *smithyhttp.Request:
+			otel.GetTextMapPropagator().Inject(ctx, propagation.HeaderCarrier(req.Header))
+		default:
+			return out, metadata, fmt.Errorf("unknown transport type %T", req)
```
Comment (Aneurysm9) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go:125]:
```suggestion
```

It doesn't seem correct to fail all other request types here.  If this instrumentation doesn't recognize the request type it seems best to simply pass it through.

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go [hunk 2]
```diff
@@ -110,6 +112,25 @@ func (m otelMiddlewares) deserializeMiddleware(stack *middleware.Stack) error {
...
 
+func (m otelMiddlewares) finalizeMiddleware(stack *middleware.Stack) error {
+	return stack.Finalize.Add(middleware.FinalizeMiddlewareFunc("OTelFinalizeMiddleware", func(
+		ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (
+		out middleware.FinalizeOutput, metadata middleware.Metadata, err error) {
+
+		// Propagate the Trace information by injecting it into the HTTP request.
+		switch req := in.Request.(type) {
+		case *smithyhttp.Request:
+			otel.GetTextMapPropagator().Inject(ctx, propagation.HeaderCarrier(req.Header))
```
Comment (Aneurysm9) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go:123]:
```suggestion
			m.propagator.Inject(ctx, propagation.HeaderCarrier(req.Header))
```

The middleware needs to take a configuration option for setting a propagator, falling back to the global propagator if none is provided.

**Review by Aneurysm9 — COMMENTED** [review]
Can you please add a CHANGELOG entry and tests for the new functionality?

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go (hunk 1 above)
Comment (lukestoward) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go]: Agreed. I have removed this.

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go (hunk 2 above)
Comment (lukestoward) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/aws.go]: Done.

**Comment by lukestoward** [comment]
@Aneurysm9 I have added tests and made the propagator configurable. Please could you re-review, thanks.

File: CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - google.golang.org/grpc/otelgrpc: Avoid getting a new Tracer for every RPC.
 - google.golang.org/grpc/otelgrpc: Deprecate Inject and Extract public funcs.
 - google.golang.org/grpc/otelgrpc: Conditionally compute message size for tracing events using proto v2 API rather than legacy v1 API.
+- Add trace propagation support to `instrumentation/github.com/aws/aws-sdk-go-v2/otelaws` (#2856).
```
Comment (Aneurysm9) [inline CHANGELOG.md:16]:
```suggestion
- Add trace context propagation support to `instrumentation/github.com/aws/aws-sdk-go-v2/otelaws` (#2856).
```

This item should go in an `### Added` heading.

**Review by Aneurysm9 — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (lukestoward) [reply CHANGELOG.md]: Done. Thanks for reviewing.

**Review by dmathieu — APPROVED** [review]

**Comment by lukestoward** [comment]
@dmathieu I've rebased and fixed the linter complaints. Please could you re-trigger the workflow. Thanks.

**Comment by lukestoward** [comment]
Awesome, looks like all checks are passing. Rather than keep rebasing, please let me know when this can be merged and I will rebase ASAP.

Please let me know if anything else is needed.

**Pull request merged**